romdisk/date.asm

```diff
@@ -49,8 +92,117 @@ date_main:
         ld bc, 20
         WRITE()
         ret
+_date_usage_error:
+        ld de, date_usage_str
+        ld bc, date_usage_str_end - date_usage_str
+        S_WRITE1(DEV_STDOUT)
+        ret
+
+        ; Read ACSII date and fill out date part of DATE_STRUCT.
+        ; The input format will be as follows:
+        ;   YYYY-MM-DD
+        ; Does not check the '-' chars, they can be anything. Does not
+        ; validate number ranges.
+        ; Parameters:
+        ;       HL - Pointer to the date structure, of size DATE_STRUCT_SIZE
+        ;       DE - String source.
+        ; Returns:
+        ;       A - nonzero if bad format detected. 
+        ;       If A is zero the date parts of the input struct have been filled out.
+        ;       DE - DE + 10
+        ; Modifies: HL, BC
+        PUBLIC ascii_to_date
+ascii_to_date:
+        xor a
+        ld b, '-'
+        call read_bcd_pair
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        ; TODO calculate and fill out day-of-week
+        ret
+
+        ; Read ACSII time and fill out time part of of DATE_STRUCT.
+        ; The input format will be as follows:
+        ;   HH:MM:SS
+        ; Does not check the ':' chars, they can be anything. Does not
+        ; validate number ranges.
+        ; Parameters:
+        ;       HL - Pointer to the time part of date structure
+        ;       DE - String source.
+        ; Returns:
+        ;       A  - nonzero if bad format detected
+        ;       The time parts of the input date struct have been filled out.
+        ; Modifies: BC
+        PUBLIC ascii_to_time
+ascii_to_time:
+        xor a
+        ld b, ':'
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        call check_char
+        call read_bcd_pair
+        ret
+
+        ; Convert an decimal ascii pair of bytes to a BCD byte.
+        ; Parameters
+        ;       DE - Pointer to first (high) ascii byte
+        ;       HL - Pointer to result byte to update
+        ; Returns:
+        ;       (HL) - 2 BCD nibbles, big endian
+        ;       HL   - HL + 1
+        ;       DE   - DE + 2
+        ;       A    - unmodified if both input chars are digit,
+        ;              nonzero otherwise
+read_bcd_pair:
+        ld c, a ; we use C as accumulated error condition
+        ld a, (de)
+        cp '9' + 1
```

Comment:
Maybe you can re-use the routine `parse_dec_digit` from `strutils.asm` file, it will help you with this

Reply:
Yes using parse_dec_digit directly (rather than copying bits) definitely improves it.